README.md
# css-shortener

[![CI](https://github.com/mbrandau/css-shortener/actions/workflows/main.yml/badge.svg)](https://github.com/mbrandau/css-shortener/actions/workflows/main.yml) [![npm](https://img.shields.io/npm/v/css-shortener.svg)](https://www.npmjs.com/package/css-shortener) [![npm](https://img.shields.io/npm/dt/css-shortener.svg)](https://www.npmjs.com/package/css-shortener) [![GitHub issues](https://img.shields.io/github/issues/mbrandau/css-shortener.svg)](https://github.com/mbrandau/css-shortener/issues)

### Preview

```css
/* BEFORE */
p.this-class-is-extremely-long-and-definitely-needs-to-be-shortened,
p.why-is-this-so-long-if-it-just-makes-white-text {
  color: white;
}

/* AFTER */
p.a,
p.b {
  color: white;
}
```

## Table of contents

1. [Quick Start](#quick-start)
2. [API Documentation](#api-documentation)
   1. [Constructor](#constructor)
      - [Options](#options)
   2. [`#importMap(map, override)`](#importmapmap-override)
   3. [`#map`](#map)
3. [Examples](#examples)
   1. [CSS filter for nunjucks and express](#css-filter-for-nunjucks-and-express)

## Quick Start

```js
import { CssShortener } from 'css-shortener';

const shortener = new CssShortener();

console.log(shortener.shortenClassName('long-class'));
// Output: a

console.log(shortener.map);
// Output: {
//   "long-class": "a"
// }
```

## API Documentation

### Constructor

```js
const options = {
  alphabet: 'abcdef',
};
const shortener = new CssShortener(options);
```

The `options` parameter can be omitted.

#### Options

| Option   | Type     | Optionality | Description                                           | Default value                             |
| -------- | -------- | ----------- | ----------------------------------------------------- | ----------------------------------------- |
| alphabet | _string_ | optional    | The alphabet is used to generate the new class names. | `'abcefghijklmnopqrstuvwxyz0123456789_-'` |

Note that there is no `d` in the default alphabet to avoid generation of the combination `ad`.

### `#importMap(map, override)`

Imports mappings into the shortener

```js
cssShortener.importMap(
  {
    'my-extremely-long-class-name': 'a',
  },
  false
);
```

If `override` is true, class names that are already mapped will be overridden.  
The `override` parameter can be omitted.

### `#map`

Returns the mapped class names

```js
var map = cssShortener.getMap();
```

```json
{
  "my-extremely-long-class-name": "a"
}
```

## Examples

### CSS filter for nunjucks and express

```js
const express = require('express');
const nunjucks = require('nunjucks');

const cssMap = JSON.parse(fs.readFileSync('./cssMap.json'));

const app = express();
const env = nunjucks.configure('your-views-folder', { express: app });

env.addFilter('css', function(str) {
  return str
    .split(' ')
    .map(c => (cssMap[c] != null ? cssMap[c] : c))
    .join(' ');
});
```

```html
<!-- BEFORE -->
<div class="{{'my-extremely-long-class-name'|css}}"></div>
<!-- RENDERED -->
<div class="a"></div>
```


CHANGELOG.md
## [3.0.0](https://github.com/mbrandau/css-shortener/compare/v2.2.0...v3.0.0) (2022-03-26)


### ⚠ BREAKING CHANGES

* trigger breaking change

### Bug Fixes

* update semantic release config ([8d04b25](https://github.com/mbrandau/css-shortener/commit/8d04b25ed86f3b531bf8359e387e2de8d3b8de7b))

# Changelog

## 3.0.0 - 26/03/2022

  - **BREAKING:** rewrite library

## 2.1.0 - 10/09/2018

  - Move generation of new class names to a public function (`getNewClassName(className)`)

## 1.2.0 - 02/25/2018

  - ADD `ignorePrefix` and `trimIgnorePrefix` options to omit classes from being replaced
  - OPTIMIZE performance for #htmlStream()
  - FIX HTML RegExp (was too greedy)
  - FIX [#2: htmlStream() only working on attributes with multiple classes](https://github.com/mbrandau/css-shortener/issues/2)
  - ADD `html` command

## 1.1.1

  - FIX [#1: Exclude comments from CSS RegExp](https://github.com/mbrandau/css-shortener/issues/1)  
  [Thanks to *revo*](https://stackoverflow.com/a/48962872/5133130) from StackOverflow

## 1.1.0

  - ADD ability to import existing mappings
  - ADD html stream to replace class names with mappings

## 1.0.0


test/index.js
const should = require('should');
const CssShortener = require('../lib/index');
const IdGenerator = require('../lib/idGenerator');
const str = require('string-to-stream');
const toString = require('stream-to-string');

describe('CssShortener', function () {
    it('should use custom alphabet', function (done) {
        const c = new CssShortener({
                alphabet: ['a']
            }),
            stream = str('.class1{}.class2{}.class3{}').pipe(c.cssStream());
        toString(stream).then(function (msg) {
            msg.should.equal('.a{}.aa{}.aaa{}');
            c.getMap().should.deepEqual({
                'class1': 'a',
                'class2': 'aa',
                'class3': 'aaa'
            });
            done();
        });
    });
    it('should use default ignorePrefix and default trimIgnorePrefix options', function (done) {
        const c = new CssShortener(),
            stream = str('.class1{}.ignore-class2{}.class3{}').pipe(c.cssStream());
        toString(stream).then(function (msg) {
            msg.should.equal('.a{}.class2{}.b{}');
            c.getMap().should.deepEqual({
                'class1': 'a',
                'class3': 'b'
            });
            done();
        });
    });
    describe('#importMap()', function () {
        it('should import mappings from a map object', function () {
            const c = new CssShortener(),
                map = {
                    'test-class': 'a'
                };
            c.importMap(map);
            c.getMap().should.deepEqual(map);
        });
        it('should not override existing mappings', function () {
            const c = new CssShortener(),
                prevMap = {
                    'test-class': 'a'
                },
                importMap = {
                    'test-class': 'b'
                };
            c.importMap(prevMap);
            c.importMap(importMap);
            c.getMap().should.deepEqual(prevMap);
        });
        it('should override existing mappings', function () {
            const c = new CssShortener(),
                prevMap = {
                    'test-class': 'a'
                },
                importMap = {
                    'test-class': 'b'
                };
            c.importMap(prevMap);
            c.importMap(importMap, true);
            c.getMap().should.deepEqual(importMap);
        });
    });
    describe('#getMap()', function () {
        it('should return a map with original class names and shortened versions', function () {
            const c = new CssShortener(),
                map = {
                    'test-class': 'a'
                };
            c.importMap(map);
            c.getMap().should.deepEqual(map);
        });
    });
    describe('#cssStream()', function () {
        it('should replace css classes with new IDs', function (done) {
            const c = new CssShortener(),
                stream = str('p.myclass{}.testclass{}').pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal('p.a{}.b{}');
                c.getMap().should.deepEqual({
                    'myclass': 'a',
                    'testclass': 'b'
                });
                done();
            });
        });
        it('should replace css classes with already mapped IDs', function (done) {
            const c = new CssShortener();
            c.importMap({
                'myclass': 'zx'
            });
            const stream = str('p.myclass{}.testclass{}').pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal('p.zx{}.a{}');
                c.getMap().should.deepEqual({
                    'myclass': 'zx',
                    'testclass': 'a'
                });
                done();
            });
        });
        it('should ignore css classes in comments', function (done) {
            const c = new CssShortener(),
                stream = str(`/* p.myclass{}.testclass{} */
/* file.css
.class
.test/*`).pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal(`/* p.myclass{}.testclass{} */
/* file.css
.class
.test/*`);
                c.getMap().should.deepEqual({});
                done();
            });
        });
        it('should replace css classes in long selectors', function (done) {
            const c = new CssShortener(),
                stream = str('p.class0,.class1,html,td.class2,tr{}.testclass{}').pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal('p.a,.b,html,td.c,tr{}.e{}');
                c.getMap().should.deepEqual({
                    'class0': 'a',
                    'class1': 'b',
                    'class2': 'c',
                    'testclass': 'e'
                });
                done();
            });
        });
        it('should ignore css classes with specified prefix', function (done) {
            const c = new CssShortener({
                    ignorePrefix: 'ignoreme-'
                }),
                stream = str('p.class0,.ignoreme-testclass{},.class1,html,td.class2,tr{}.testclass{}').pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal('p.a,.testclass{},.b,html,td.c,tr{}.e{}');
                c.getMap().should.deepEqual({
                    'class0': 'a',
                    'class1': 'b',
                    'class2': 'c',
                    'testclass': 'e'
                });
                done();
            });
        });
        it('should not trim ignorePrefix if trimIgnorePrefix is false', function (done) {
            const c = new CssShortener({
                    trimIgnorePrefix: false
                }),
                stream = str('p.class0,.ignore-testclass{},.class1,html,td.class2,tr{}.testclass{}').pipe(c.cssStream());
            toString(stream).then(function (msg) {
                msg.should.equal('p.a,.ignore-testclass{},.b,html,td.c,tr{}.e{}');
                c.getMap().should.deepEqual({
                    'class0': 'a',
                    'class1': 'b',
                    'class2': 'c',
                    'testclass': 'e'
                });
                done();
            });
        });
    });
    describe('#replaceCss()', function () {
        it('should replace css classes with new IDs', function () {
            const c = new CssShortener();
            c.replaceCss('p.myclass{}.testclass{}').should.equal('p.a{}.b{}');
            c.getMap().should.deepEqual({
                'myclass': 'a',
                'testclass': 'b'
            });
        });
        it('should replace css classes with already mapped IDs', function () {
            const c = new CssShortener();
            c.importMap({
                'myclass': 'zx'
            });
            c.replaceCss('p.myclass{}.testclass{}').should.equal('p.zx{}.a{}');
            c.getMap().should.deepEqual({
                'myclass': 'zx',
                'testclass': 'a'
            });
        });
        it('should ignore css classes in comments', function () {
            const c = new CssShortener();
            c.replaceCss(`/* p.myclass{}.testclass{} */
/* file.css
.class
.test/*`).should.equal(`/* p.myclass{}.testclass{} */
/* file.css
.class
.test/*`);
            c.getMap().should.deepEqual({});
        });
        it('should replace css classes in long selectors', function () {
            const c = new CssShortener();
            c.replaceCss('p.class0,.class1,html,td.class2,tr{}.testclass{}').should.equal('p.a,.b,html,td.c,tr{}.e{}');
            c.getMap().should.deepEqual({
                'class0': 'a',
                'class1': 'b',
                'class2': 'c',
                'testclass': 'e'
            });
        });
        it('should ignore css classes with specified prefix', function () {
            const c = new CssShortener({
                ignorePrefix: 'ignoreme-'
            });
            c.replaceCss('p.class0,.ignoreme-testclass{},.class1,html,td.class2,tr{}.testclass{}').should.equal('p.a,.testclass{},.b,html,td.c,tr{}.e{}');
            c.getMap().should.deepEqual({
                'class0': 'a',
                'class1': 'b',
                'class2': 'c',
                'testclass': 'e'
            });
        });
        it('should not trim ignorePrefix if trimIgnorePrefix is false', function () {
            const c = new CssShortener({
                trimIgnorePrefix: false
            });
            c.replaceCss('p.class0,.ignore-testclass{},.class1,html,td.class2,tr{}.testclass{}').should.equal('p.a,.ignore-testclass{},.b,html,td.c,tr{}.e{}');
            c.getMap().should.deepEqual({
                'class0': 'a',
                'class1': 'b',
                'class2': 'c',
                'testclass': 'e'
            });
        });
    });
    describe('#htmlStream()', function () {
        it('should replace mapped css classes', function (done) {
            const c = new CssShortener(),
                map = {
                    'test-class': 'a'
                };
            c.importMap(map);

            const stream = str('<div class="abc test-class 15a"></div>').pipe(c.htmlStream());
            toString(stream).then(function (msg) {
                msg.should.equal('<div class="abc a 15a"></div>');
                const stream = str('<div class="test-class"></div>').pipe(c.htmlStream());
                toString(stream).then(function (msg) {
                    msg.should.equal('<div class="a"></div>');
                    const stream = str('<div class="myclass"></div>').pipe(c.htmlStream());
                    toString(stream).then(function (msg) {
                        msg.should.equal('<div class="myclass"></div>');
                        done();
                    });
                });
            });
        });
    });
    describe('#replaceHtml()', function () {
        it('should replace mapped css classes', function () {
            const c = new CssShortener(),
                map = {
                    'test-class': 'a'
                };
            c.importMap(map);

            const html = '<div class="abc test-class 15a"></div>';
            const replaced = c.replaceHtml(html);
            replaced.should.equal('<div class="abc a 15a"></div>');

            c.replaceHtml('<div class="test-class"></div>').should.equal('<div class="a"></div>');

            c.replaceHtml('<div class="myclass"></div>').should.equal('<div class="myclass"></div>');
        });
    });
})
;

describe('IdGenerator (default alphabet)', function () {
    const gen = new IdGenerator();
    it('should not start with a number or a hyphen', function () {
        for (let i = 0; i < 100; i++) {
            gen().should.not.match(/^[0-9-].*$/);
        }
    });
});


test/index.test.ts
import { CssShortener } from '../src';
import { createIdGenerator } from '../src/idGenerator';

describe('CssShortener', function() {
  it('should use custom alphabet', function() {
    const c = new CssShortener({
      alphabet: 'a',
    });
    expect(c.shortenClassName('class1')).toBe('a');
    expect(c.shortenClassName('class2')).toBe('aa');
    expect(c.shortenClassName('class3')).toBe('aaa');
    expect(c.map).toStrictEqual({
      class1: 'a',
      class2: 'aa',
      class3: 'aaa',
    });
  });
  describe('#importMap()', function() {
    it('should import mappings from a map object', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'a',
        };
      c.importMap(map);
      expect(c.map).toStrictEqual(map);
    });
    it('should not override existing mappings', function() {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class': 'b',
        };
      c.importMap(prevMap);
      c.importMap(importMap);
      expect(c.map).toStrictEqual(prevMap);
    });
    it('should override existing mappings', function() {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class': 'b',
        };
      c.importMap(prevMap);
      c.importMap(importMap, true);
      expect(c.map).toStrictEqual(importMap);
    });
    it('throws when a short name to import is already in the map', () => {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class2': 'a',
        };
      c.importMap(prevMap);
      expect(() => c.importMap(importMap)).toThrowErrorMatchingSnapshot();
    });
  });
  describe('#map', function() {
    it('should return a map with original class names and shortened versions', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'a',
        };
      c.importMap(map);
      expect(c.map).toStrictEqual(map);
    });
  });
  describe('#shortenClassName', function() {
    it('should not reuse already existing short class names', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'b',
        };
      c.importMap(map);
      expect(c.shortenClassName('test1')).toBe('a');
      expect(c.shortenClassName('test2')).toBe('c');
      expect(c.shortenClassName('test-class')).toBe('b');
    });
  });
});

describe('IdGenerator (default alphabet)', function() {
  const gen = createIdGenerator();
  it('should not start with a number or a hyphen', function() {
    for (let i = 0; i < 100; i++) {
      expect(gen()).not.toMatch(/^[0-9-].*$/);
    }
  });
});


src/index.ts
import { createIdGenerator } from './idGenerator';

export interface Options {
  /**
   * Alphabet to use for generating ids.
   */
  alphabet?: string;
}

export type ClassNameMap = { [longName: string]: string };

export class CssShortener {
  private _map: ClassNameMap = {};
  private generateId: () => string;

  constructor(options?: Options) {
    this.generateId = createIdGenerator(options?.alphabet);
  }

  public get map(): ClassNameMap {
    return this._map;
  }

  /**
   * Imports mapped CSS class names
   * @param {object} map Map that should be imported
   * @param {boolean} override If true, existing mappings will be overridden
   */
  public importMap(map: ClassNameMap, override: boolean = false) {
    for (let classNameToImport in map) {
      if (this.isAlreadyUsed(map[classNameToImport]))
        throw new Error(
          `Cannot import class '${classNameToImport}'. Short class name '${map[classNameToImport]}' is already being used.`
        );
      if (this._map[classNameToImport] != null) {
        // Override mapped class name
        if (override === true)
          this._map[classNameToImport] = map[classNameToImport];
      } else this._map[classNameToImport] = map[classNameToImport]; // Import class name
    }
  }

  /**
   * Checks if the given class name is already used as a replacement for a longer class name
   * @param className class name to check for
   */
  private isAlreadyUsed(className: string): boolean {
    return Object.values(this.map).includes(className);
  }

  /**
   * Generates and maps a new class name
   * @param className The existing class name that should be replaced
   * @returns {string} The new generated and mapped class name
   */
  public shortenClassName(className: string) {
    // Return the already mapped class name
    if (this._map[className] != null) return this._map[className];
    // Generate, map and return the new class name
    let newClassName = this.generateId();
    while (this.isAlreadyUsed(newClassName)) {
      newClassName = this.generateId();
    }
    return (this._map[className] = newClassName);
  }
}


src/idGenerator.js
// Leaving 'd' out to avoid generating the word 'ad'
const defaultAlphabet = 'abcefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-';

var IdGenerator = function(alphabet) {
  var options = {
    alphabet: alphabet || defaultAlphabet,
    length: 1,
    index: 0
  };

  return function() {
    var res = generateId(options);
    while (/^[0-9-].*$/.test(res)) res = generateId(options);
    return res;
  };
};

var generateId = function(options) {
  var res = '';

  for (var i = options.length - 1; i >= 0; i--) {
    var x = Math.pow(options.alphabet.length, i);
    var n = Math.floor(options.index / x);
    res += options.alphabet[n % options.alphabet.length];
  }

  options.index++;
  if (options.index > Math.pow(options.alphabet.length, options.length) - 1) {
    options.length++;
    options.index = 0;
  }

  return res;
};

module.exports = IdGenerator;


src/idGenerator.ts
// Leaving 'd' out to avoid generating the word 'ad'
const defaultAlphabet = 'abcefghijklmnopqrstuvwxyz0123456789_-';

export function createIdGenerator(alphabet?: string) {
  var options = {
    alphabet: alphabet || defaultAlphabet,
    length: 1,
    index: 0,
  };

  return function() {
    var res = generateId(options);
    // class names should not start with a number or a dash
    while (/^[0-9-].*$/.test(res)) res = generateId(options);
    return res;
  };
}

var generateId = function(options: {
  length: number;
  alphabet: string;
  index: number;
}) {
  var res = '';

  for (var i = options.length - 1; i >= 0; i--) {
    var x = Math.pow(options.alphabet.length, i);
    var n = Math.floor(options.index / x);
    res += options.alphabet[n % options.alphabet.length];
  }

  options.index++;
  if (options.index > Math.pow(options.alphabet.length, options.length) - 1) {
    options.length++;
    options.index = 0;
  }

  return res;
};


src/index.js
const replaceStream = require('replacestream');
const IdGenerator = require('./idGenerator');

const CLASS_NAME_REGEX = /\/\*[\s\S]*?\*\/|(\.[a-z_-][\w-]*)(?=[^{}]*{)/g; // https://stackoverflow.com/a/48962872/5133130
const HTML_CLASS_REGEX = /class=["|']{1}([\w-\s]*)["|']{1}/g;

/**
 * Provides methods to replaces CSS class names
 * @constructor
 */
const CssShortener = function(options) {
  if (!options) options = {};
  this._options = options;
  if (!this._options.hasOwnProperty('ignorePrefix'))
    this._options.ignorePrefix = 'ignore-';
  if (!this._options.hasOwnProperty('trimIgnorePrefix'))
    this._options.trimIgnorePrefix = true;
  this._idGenerator = new IdGenerator(this._options.alphabet);
  this._classNameMap = {};

  const t = this;
  const replaceCss = function(match, capturingGroup) {
    if (!capturingGroup) return match;
    const className = capturingGroup.substr(1); // Remove dot infront of class name

    return `.${t.getNewClassName(className)}`;
  };
  const replaceHtml = function(match, capturingGroup) {
    if (!capturingGroup) return match;
    const classes = capturingGroup.trim().split(' ');
    const classCount = classes.length;
    let result = '';
    for (let i = 0; i < classCount; i++) {
      // Check if class is mapped and add it to the result
      result +=
        t._classNameMap[classes[i]] != null
          ? t._classNameMap[classes[i]]
          : classes[i];
      if (i < classCount - 1) result += ' ';
    }
    return `class="${result}"`;
  };

  /**
   * Returns all mapped CSS class names
   * @returns {object}
   */
  this.getMap = function() {
    return this._classNameMap;
  };

  /**
   * Imports mapped CSS class names
   * @param {object} map Map that should be imported
   * @param {boolean} override If true, existing mappings will be overridden
   */
  this.importMap = function(map, override) {
    for (let classNameToImport in map) {
      if (this._classNameMap[classNameToImport] != null) {
        // Override mapped class name
        if (override === true)
          this._classNameMap[classNameToImport] = map[classNameToImport];
      } else this._classNameMap[classNameToImport] = map[classNameToImport]; // Import class name
    }
  };

  /**
   * Generates and maps a new class name
   * @param className The existing class name that should be replaced
   * @returns {string} The new generated and mapped class name
   */
  this.getNewClassName = function(className) {
    // If the ignorePrefix option is set and the current class starts with the prefix, trim the prefix off and ignore the class.
    if (
      t._options.ignorePrefix &&
      className.startsWith(t._options.ignorePrefix)
    )
      return t._options.trimIgnorePrefix
        ? className.substr(t._options.ignorePrefix.length)
        : className;

    // Return the already mapped class name
    if (t._classNameMap[className] != null) return t._classNameMap[className];
    // Generate, map and return the new class name
    else return (t._classNameMap[className] = t._idGenerator());
  };

  /**
   * @returns {Transform} A transform stream that replaces class names in the CSS code
   */
  this.cssStream = function() {
    return replaceStream(CLASS_NAME_REGEX, replaceCss);
  };

  /**
   * Replaces class names in the given CSS code.
   * @param css The CSS code in which class names will be replaced
   * @returns {string}
   */
  this.replaceCss = function(css) {
    return css.replace(CLASS_NAME_REGEX, replaceCss);
  };

  /**
   * Class names will only be replaced with the existing mappings. No mappings will be generated.
   * @returns {Transform} A transform stream that replaces class names in the HTML code
   */
  this.htmlStream = function() {
    return replaceStream(HTML_CLASS_REGEX, replaceHtml);
  };

  /**
   * Replaces class names in the given HTML code.
   * Class names will only be replaced with the existing mappings. No mappings will be generated.
   * @param html The HTML code in which class names will be replaced
   * @returns {string}
   */
  this.replaceHtml = function(html) {
    return html.replace(HTML_CLASS_REGEX, replaceHtml);
  };
};

module.exports = CssShortener;
